argon2/src/memory.rs

```diff
@@ -0,0 +1,210 @@
+//! Views into Argon2 memory that can be processed in parallel.
+//!
+//! This module implements, with a combination of compile-time borrowing and runtime checking, the
+//! cooperative contract described in section 3.4 (Indexing) of RFC 9106:
+//!
+//! > To enable parallel block computation, we further partition the memory matrix into SL = 4
+//! > vertical slices. The intersection of a slice and a lane is called a segment, which has a
+//! > length of q/SL. Segments of the same slice can be computed in parallel and do not reference
+//! > blocks from each other. All other blocks can be referenced.
+
+#![warn(
+    clippy::undocumented_unsafe_blocks,
+    clippy::missing_safety_doc,
+    unsafe_op_in_unsafe_fn
+)]
+
+use core::marker::PhantomData;
+use core::ptr::NonNull;
+
+#[cfg(feature = "parallel")]
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+
+use crate::{Block, SYNC_POINTS};
+
+pub trait SegmentViews<'a> {
+    /// Construct an iterator of parallelizable views into a set of Argon2 memory blocks.
+    ///
+    /// If the `parallel` feature is enabled, the returned type implements
+    /// [`rayon::iter::ParallelIterator`]; otherwise it implements [`core::iter::Iterator`].
+    fn segment_views(&mut self, slice: usize, lanes: usize) -> SegmentViewIter<'_>;
+}
+
+impl<'a> SegmentViews<'a> for &'a mut [Block] {
+    fn segment_views(&mut self, slice: usize, lanes: usize) -> SegmentViewIter<'_> {
+        // SAFETY: a pointer to the buffer or a `&mut [Block]` can never be null.
+        let blocks = unsafe { NonNull::new_unchecked(self.as_mut_ptr()) };
```

Comment:
[`NonNull:from`](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html#impl-From%3C%26mut+T%3E-for-NonNull%3CT%3E) works on mutable references, so a possible reduction in `unsafe` noise.

Reply:
Hey, thanks for the review!

One issue is that `NonNull::from(self)` would compile fine... even if someone later changed the method to just take `&self`, resulting in undefined behavior (because we would mutate through a pointer derived from a shared reference, which is UB).

That's why I opted to not rely on the `From<&mut T>` impl. But maybe I should document that in the code. Or, alternatively, I could use the impl but leave a warning that the pointer MUST be derived from a mutable reference. I think the former is harder to accidentally ignore, but I'm open to different opinions/preferences.

Reply:
We also want a `NonNull<Block>` pointer, not a `NonNull<&mut [Block]>`.

---

Regardless, there's a safe but still explicit alternative:

```suggestion
        // Note that this pointer needs to be derived from a mutable slice/reference, otherwise
        // mutating through it later would be UB.
        let blocks = NonNull::new(self.as_mut_ptr()).expect("a reference cannot dangle");
```

A quick test on Compiler Explorer shows that Rust should indeed remove the unnecessary check and panic code. So this should result in identical code as the version using `NonNull::new_unchecked`.

Reply:
Nice mini investigation :D

Shame https://github.com/rust-lang/rust/issues/130823 didn't stabilize sooner but `.expect()` seems like a better middle for a security sensitive crate.